### DISCUSSION TITLE
[Feature] use linux Nice to avoid fighting ffmpeg for CPU time

### DISCUSSION THREAD

**Discussion by awoimbee** [immich-app/immich#1970] [Feature Request]
### The feature

immich server is made of lots of components, the API needs to respond fast while ffmpeg and ML services can run in the background.
The issue is that these background services are using 100% CPU (on my rpi4) and the API has to fight for CPU time.
A good solution would be to use [Nice](https://en.wikipedia.org/wiki/Nice_(Unix)). API should be level -1 or -2 while ffmpeg can safely go to 19.

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by EnochPrime** [comment]
An alternative that could be done on the user end right now is to add resource limits to the machine learning container. To do this on the pi you'll need to enable cgroups.

https://docs.docker.com/compose/compose-file/compose-file-v3/#resources

However, this does mean that it is always limited so you can't take advantage of less active times.

**Reply by bo0tzz** [reply]
If you set a cpu reservation but no limit, I think that should behave as desired - where the container gets to use full cpu when available, but is kept back to the level of the reservation when other things need the cpu.

**Reply by awoimbee** [reply]
Good idea ! This seems way simpler and more platform agnostic.
I think a better feature request would be "add resources reservations to [docker-compose](https://github.com/immich-app/immich/blob/main/docker/docker-compose.yml)"

EDIT: It can't work for ffmpeg (which is the biggest CPU hog for me) because ffmpeg runs inside the `immich_server` container.

**Reply by bo0tzz** [reply]
> add resources reservations to docker-compose

I've considered this before, but people's resource availability is so varied that there probably isn't a good default.

> ffmpeg runs inside the immich_server container

`server` and `microservices` use the same image (which contains ffmpeg), but ffmpeg should only be running in the `microservices` container. You can cap the CPU on that while leaving `server` unrestrained.

**Reply by bigntallmike** [reply]
> `server` and `microservices` use the same image (which contains ffmpeg), but ffmpeg should only be running in the `microservices` container. You can cap the CPU on that while leaving `server` unrestrained.

Sorry for the necro, but could someone please post an example of how and where this would be specified? I don't see any reference to `microservices` in `docker-compose.yml` and the documentation doesn't really clear that up for me.

**Reply by bo0tzz** [reply]
This post is very old, it's now consolidated into one container.
